runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp

```diff
@@ -0,0 +1,1434 @@
+	*signatureLength = 2; //space for '(', ')'
+	for (U_32 i = 0; i < numArgs; i++) {
+		j9object_t pObject = J9JAVAARRAYOFOBJECT_LOAD(currentThread, ptypes, i);
+		J9Class *pclass = J9VM_J9CLASS_FROM_HEAPCLASS(currentThread, pObject);
+		signatures[i] = sigForPrimitiveOrVoid(vm, pclass);
+		if (!signatures[i])
+			signatures[i] = getClassSignature(currentThread, pclass);
+
+		*signatureLength += strlen(signatures[i]);
+	}
+
+	// Return type
+	j9object_t rtype = J9VMJAVALANGINVOKEMETHODTYPE_RTYPE(currentThread, typeObject);
+	J9Class *rclass = J9VM_J9CLASS_FROM_HEAPCLASS(currentThread, rtype);
+	char* rSignature = sigForPrimitiveOrVoid(vm, rclass);
+	if (!rSignature)
+		rSignature = getClassSignature(currentThread, rclass);
+
+	*signatureLength += strlen(rSignature);
+
+	char* methodDescriptor = (char*)j9mem_allocate_memory(*signatureLength+1, OMRMEM_CATEGORY_VM);
+	char* cursor = methodDescriptor;
+	*cursor++ = '(';
+
+	// Copy class signatures to descriptor string
+	for (U_32 i = 0; i < numArgs; i++) {
+		U_32 len = strlen(signatures[i]);
+		strncpy(cursor, signatures[i], len);
+		cursor += len;
+		j9mem_free_memory(signatures[i]);
+	}
+
+	*cursor++ = ')';
+	// Copy return type signature to descriptor string
+	strcpy(cursor, rSignature);
+	j9mem_free_memory(signatures);
+	return methodDescriptor;
+}
+
+static J9Class *
+classForSignature(struct J9VMThread *vmThread, U_8 **sigDataPtr, struct J9ClassLoader *classLoader)
+{
+	U_8 *sigData = *sigDataPtr;
+	J9JavaVM *vm = vmThread->javaVM;
+	J9Class *clazz = NULL;
+	UDATA arity = 0;
+	U_8 c = 0;
+	UDATA i = 0;
+
+	for (c = *sigData++; '[' == c; c = *sigData++) {
+		arity++;
+	}
+
+	/* Non-array case */
+	switch (c) {
+	case 'Q':
+	case 'L': {
+		/* object case */
+		U_8 *tempData = sigData;
+		UDATA length = 0;
+
+		/* find the signature length -> up to the ";" */
+		while (';' != *sigData++) {
+			length++;
+		}
+
+		/* find the class from the signature chunk */
+		clazz = vm->internalVMFunctions->internalFindClassUTF8(vmThread, tempData, length, classLoader, J9_FINDCLASS_FLAG_THROW_ON_FAIL);
+		break;
+	}
+	case 'I':
+		clazz = vm->intReflectClass;
+		break;
+	case 'Z':
+		clazz = vm->booleanReflectClass;
+		break;
+	case 'J':
+		clazz = vm->longReflectClass;
+		break;
+#if defined(J9VM_INTERP_FLOAT_SUPPORT)
+	case 'D':
+		clazz = vm->doubleReflectClass;
+		break;
+	case 'F':
+		clazz = vm->floatReflectClass;
+		break;
+#endif
+	case 'C':
+		clazz = vm->charReflectClass;
+		break;
+	case 'B':
+		clazz = vm->byteReflectClass;
+		break;
+	case 'S':
+		clazz = vm->shortReflectClass;
+		break;
+	case 'V':
+		clazz = vm->voidReflectClass;
+		break;
+	}
+
+	for (i = 0; (i < arity) && (NULL != clazz); i++) {
+		clazz = fetchArrayClass(vmThread, clazz);
+	}
+
+	if (NULL != clazz) {
+		*sigDataPtr = sigData;
+	}
+
+	return clazz;
+}
+
+static j9object_t
+createFieldFromID(struct J9VMThread *vmThread, J9JNIFieldID *j9FieldID)
```

Comment:
was this duplicated from the reflectHelpers code?  Any reason to make another copy of it?

Reply:
I wasn't sure if I should expand the reflection function table to export this API as this will be the single place to use it outside of reflecthelp.c, so I made the copy.

Reply:
Better to expand the table than to have to maintain a duplicate of the code.  Otherwise, it's easy to miss updating one of the two places if we find a bug